### PR TITLE
Re-use cancellation tokens in the https middleware

### PR DESCRIPTION
**Pull request by davidfowl**
- These tokens are usually short lived and exist for the purposes of cancelling the handshake.
- The implementation doesn't have a dispose. The cost of storing 1024 tokens is tiny and it'll be unrooted once the server is unrooted.

PS: The pool should remain mostly small in the normal case when handshakes are fast. There will be allocations when there are delays in the handshake (more then 1024) handshakes happening concurrently. Returning to pool takes a timer queue lock here so we'll need to run this without handshake test to see if there are any regressions.

cc @stephentoub

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+using System.Collections.Concurrent;
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs:1]: nit: copyright

File: src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
+{
+    internal class CancellationTokenSourcePool
```
Comment (jkotalik) [inline src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs:6]: Shouldn't this be disposable and dispose all tokens in the queue?

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs:6]: I put a note about it in the PR description.

> The implementation doesn't have a dispose. The cost of storing 1024 tokens is tiny and it'll be unrooted once the server is unrooted.

**Review by jkotalik — APPROVED** [review]

**Comment by halter73** [comment]
As straightforward and similar to other pools as this is, I'm not sure it's worth the extra complexity to use it in HttpsConnectionMiddleware. We don't really do cross-connection pooling in Kestrel outside of memory buffers.

In the context of the SslStream allocation and the allocations involved in AuthenticateAsServerAsync, I'm not sure it's worth it. Are we going to start pooling _all_ our CancellationTokenSources? I'd feel very different if this was per-request.

File: src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                Interlocked.Decrement(ref _count);
+                return cts;
+            }
+            return new PooledCancellationTokenSource(this);
+        }
+
+        private bool Return(PooledCancellationTokenSource cts)
+        {
+            // This counting isn't accurate, but it's good enough for what we need to avoid using _queue.Count which could be expensive
+            if (!cts.TryReset() || Interlocked.Increment(ref _count) > MaxQueueSize)
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs:29]: We shouldn't call `Interlocked.Decrement(ref _count)` when `!cts.TryReset()`. That risks making the count negative.

```suggestion
            if (Interlocked.Increment(ref _count) > MaxQueueSize || !cts.TryReset())
```

Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs:29]: I don't love the "This counting isn't accurate," comment either. It's accurate enough to ensure `_queue` will never have more than `MaxQueueSize` items at any given time.

**Comment by davidfowl** [comment]
> In the context of the SslStream allocation and the allocations involved in AuthenticateAsServerAsync, I'm not sure it's worth it. Are we going to start pooling all our CancellationTokenSources? I'd feel very different if this was per-request.

We're going to pool anything that has a timeout and is short lived. No matter if it's per connection or per request.

File: src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs (hunk 3 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/CancellationTokenSourcePool.cs]: Addressed the feedback, nice catch on the count.

**Review by halter73 — APPROVED** [review]

**Pull request merged**